shared/util.c

```diff
@@ -387,9 +387,13 @@ static inline int is_dir(const char *path)
 
 int mkdir_p(const char *path, int len, mode_t mode)
 {
-	char *start, *end;
+	_cleanup_free_ char *start = strdup(path);
+	char *end;
 
-	start = strndupa(path, len);
+	if (start == NULL)
+		return -ENOMEM;
+
+	start[len] = '\0';
```

Comment:
so now you depend on len being >= strlen(path), otherwise you are writing past the allocation. I don't think this change is actually required and we'd better leave this to its own PR.

Anyway, better alternative here would be do: 

```
start = memdup(path, len + 1);
if (!start)
        return -ENOMEM;

start[len] = '\0';
```

Reply:
I'm ok with dropping it as indeed no call to mkdir_p is in a hot path and this already caused too many problems. If we need it in the future, we can copy strndupa_safe() from systemd and forget getting that from libc.

Let's just used memdup instead for now.

Reply:
shared/util.h:void *memdup(const void *p, size_t n) __attribute__((nonnull(1)));